JBPM-5381: Stunner & Guvnor & WB commons integration.

romartin opened the pull request:
Hey @manstis , @tsurdilo , @jlindop !
Here is the Stunner integration with KIE workbench & guvnor commons screens, editors and services.

See -> https://issues.jboss.org/browse/JBPM-5381
Depends on -> https://github.com/droolsjbpm/droolsjbpm-build-bootstrap/pull/302

Some updates:
- Refactors, improvements and fixings on backend side, which was really still a "proptotype".
- Added all workbench & guvnor integration stuff into children modules of kie-wb-common-stunner-project
- Added the concrete resource type, handlers and editors for BPMN - module kie-wb-common-bpmn-project
- Added another showcase that shows the integration stuff - module kie-wb-common-stunner-showcase-project
- Current existing showcase still present, as it loads double faster and consumes half of memory compared to the project showcase, and renamed to kie-wb-common-stunner-project-standalone
- Removed kie-wb-common-stunner-workbench module, the screens presents on that module has been moved to the standalone showcase.
- Added some tests on different modules
- Refactored the BPMN file suffix to "bpmn2" everywhere as the current jbpm-designer does. Refactored suffix on all files on the project as well.

Still improving and adding new tests and doing bug fixing, but the integration and showcases I are ready enough for merging and letting others to continue with the work (depends on this). Thanks!

romartin commented:
oopps, trying to reproduce and fix locally.

romartin commented:
Added new commit that fix the Jenkins build, sorry it was missing a source artifact. Rebuilding

romartin commented:
Green again :)

tsurdilo commented:
@romartin this looks good to me. I can run the tests from your branch. 

So +1 just one question - 
I don't see where the association to a asset type definitions is (or will it be added (for example  BpmnTypeDefinition implements ResourceTypeDefinition { ... }) - thanks!

romartin commented:
Hey @tsurdilo ! Good! Btw the asset definition is inside the bpmn modules - >https://github.com/romartin/kie-wb-common/tree/stunner-project-integration/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-project

romartin commented:
Hey guys, as this ticket has not been merged yet, just added right now a new commit for fixing an issue.  Jeknins should be building again. Thanks!

romartin commented:
Rebased with master again to avoid Jenkins complains. 
Jenkins please retest this.

romartin commented:
Hi, as not merged yet, I have added some minor fixes and renamed again the use of bpmn2 extensions for Stunner to just "bpmn", doing so allows the use of both process designers on the kie-wb as they're become different resource types. 
Rebuilding... Thanks!

romartin commented:
Hi @csadilek , just realized that @manstis is on PTO this week.. please could you take a look at this one?

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/AbstractDefinitionSetServices.java:
```diff
@@ -0,0 +1,49 @@
+import org.kie.workbench.common.stunner.core.diagram.Metadata;
+import org.kie.workbench.common.stunner.core.graph.Graph;
+
+public abstract class AbstractDefinitionSetServices implements DefinitionSetServices {
```
DefinitionSetService is a better name, why plural?

romartin replied:
agree :)

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/AbstractDefinitionSetServices.java:
```diff
@@ -0,0 +1,49 @@
+    private final DiagramMarshaller<Graph, Metadata, Diagram<Graph, Metadata>> diagramMarshaller;
+
+    protected AbstractDefinitionSetServices() {
+        this( null );
```
Consider using Java 8 Optional

romartin replied:
not following you sorry... what you mean? about the AbstractDefinitionSetServices class?

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/AbstractDiagramLookupService.java:
```diff
@@ -0,0 +1,101 @@
+            Map<String, String> criteriaMap = AbstractCriteriaLookupManager.parseCriteria( criteria );
+            String pathRaw = criteriaMap.get( "path" );
+            if ( !isEmpty( pathRaw ) ) {
+                // TODO
```
TODO what?

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/AbstractDiagramLookupService.java:
```diff
@@ -0,0 +1,101 @@
+        return diagramService;
+    }
+
+    protected boolean isEmpty( String s ) {
```
not needed StringUtils are on the classpath.

romartin replied:
oh sure!

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/DiagramLookupServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+        }
+        String m = "Criteria [" + criteria + "] not supported.";
+        LOG.error( m );
+        throw new UnsupportedOperationException( m );
```
That's not the right exception type, as clearly the operation is supported.

romartin replied:
agree :)

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/DiagramServiceImpl.java:
```diff
@@ -0,0 +1,289 @@
+    private org.uberfire.java.nio.file.Path root;
+
+    protected DiagramServiceImpl() {
+        this( null, null, null, null, null, null );
```
eeeek is this for testing purposes?

romartin replied:
No! There are two showcases named standalone & project ( integrates with guvnor and kie wb commons etc ). The service when uisng the standalone showcase is this Impl one, there is no project information, so it basically creates a stunner repository when starting up for first time and deploy some examples. 
On the other hand, when using the project integration artifacts, the service resolved is the concrete implementation named ProjectDiagramService... 
Basically the standalone version help us to be faster on the deverloment cycle, as the app runs twice faster, and ensures the architecture is still usable outside the workbench project environment, as Mark expects for other stuff...

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/DiagramServiceImpl.java:
```diff
@@ -0,0 +1,289 @@
+        // Load vfs diagrams and put into the parent registry.
+        final Collection<Diagram> diagrams = getAllDiagrams();
+        if ( null != diagrams && !diagrams.isEmpty() ) {
+            for ( Diagram diagram : diagrams ) {
```
start using Java8. diagrams.forEach...

romartin replied:
good point

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/DiagramServiceImpl.java:
```diff
@@ -0,0 +1,289 @@
+
+        } catch ( org.uberfire.java.nio.file.NoSuchFileException e ) {
+
+            return null;
```
seems worth throwing an exception if the file isn't found, no?

romartin replied:
good point

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/DiagramServiceImpl.java:
```diff
@@ -0,0 +1,289 @@
+                            // Register it into VFS storage.
+                            registerIntoVFS( f );
+                            // Delete file after added into app's vfs.
+                            // f.delete();
```
delete commented out code or uncomment in case it's needed.

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/DiagramServiceImpl.java:
```diff
@@ -0,0 +1,289 @@
+    private boolean isFileNameAccepted( String name ) {
+        if ( name != null && name.trim().length() > 0 ) {
+            Set<String> exts = getExtensionsAccepted();
+            for ( String ext : exts ) {
```
use Java8 stream anyMatch() instead of this loop.

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/main/java/org/kie/workbench/common/stunner/backend/service/XMLEncoderDiagramMetadataMarshaller.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public Metadata unmarshall( InputStream input ) throws IOException {
+        XMLDecoder decoder= new XMLDecoder( input );
```
code formatting.

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-backend/src/test/java/org/kie/workbench/common/stunner/backend/service/ErraitDiagramMetadataMarshallerTest.java:
```diff
@@ -0,0 +1,57 @@
+
+// TODO: @RunWith( MockitoJUnitRunner.class )
+@Ignore
+public class ErraitDiagramMetadataMarshallerTest {
```
Type in class name.

csadilek replied:
@romartin want to fix the typo in the class name ErraiTDiagram or delete this class since its on ```@Ignore```?

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/session/presenter/impl/AbstractClientSessionPresenter.java:
```diff
@@ -0,0 +1,131 @@
+public abstract class AbstractClientSessionPresenter<S extends AbstractClientSession, V extends ClientSessionPresenter.View>
+        implements ClientSessionPresenter<AbstractCanvas, AbstractCanvasHandler, S, V> {
+
+    // ClientSessionPresenter<AbstractCanvas, AbstractCanvasHandler, ClientSession<AbstractCanvas, AbstractCanvasHandler>, ClientSessionPresenter.View>
```
remove

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/toolbar/command/AbstractToolbarSessionCommand.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    protected String getConfirmMessage() {
+        return "Are you sure?";
```
//TODO i18n , doesn't have to be now but worth noting in source.

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/toolbar/command/AbstractToolbarSessionCommand.java:
```diff
@@ -0,0 +1,177 @@
+
+
+
+
```
The big void....

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-backend-api/src/main/java/org/kie/workbench/common/stunner/core/definition/service/DiagramMarshaller.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+
```
javadocs.

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-backend-api/src/main/java/org/kie/workbench/common/stunner/core/definition/service/DiagramMetadataMarshaller.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.IOException;
+import java.io.InputStream;
+
+public interface DiagramMetadataMarshaller<M extends Metadata> {
```
javadocs.

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/factory/diagram/DiagramFactory.java:
```diff
@@ -0,0 +1,33 @@
+ */
+public interface DiagramFactory extends Factory<Diagram, String> {
+
+    Diagram build( String name );
```
javaocs.

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/service/ClientDiagramServices.java:
```diff
@@ -17,91 +17,29 @@
-    DefinitionManager definitionManager;
-    Caller<DiagramService> diagramServiceCaller;
-    Caller<DiagramLookupService> diagramLookupServiceCaller;
+public class ClientDiagramServices extends AbstractClientDiagramServices<Diagram, DiagramService> {
```
service not services....

romartin replied:
well on client side I use "services" in plural as this class can provide shortcuts for different remote services, as currently do, it provides methods for both diagram and lookup services. But yeah, minor thing, just letting you that there was a reason hehe :) Anyway, let's be homogenious and use the singular, np! Thanks man!

csadilek commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/util/StunnerClientLogger.java:
```diff
@@ -32,4 +53,154 @@ public static String getErrorMessage( final ClientRuntimeError error ) {
+        log( "Canvas command mgr = "  + ( null != canvasCommandManager ? canvasCommandManager.toString() : "null" ) );
+    }
+
+        @SuppressWarnings( "unchecked" )
```
code formatting (please check for all files)

romartin commented:
Thanks @csadilek ! Absolutely agree on the size of this PR, sorry.. .will take into account for further times providing smaller ones :) Agree on all changes,code formattings, etc, will provide a new commit hope in a bit. Thanks again!

romartin commented:
Hi @csadilek ,
Thanks for the review! Some updates:
- Review of the different comments above
- Refactored as much as possible some collections usages to the java8 streams
- Added javadocs of some api and important interfaces and classes
- Code formatting
- Added all found i18n todos
- Rebased with master again
Thanks!!

romartin commented:
Hi @csadilek , just added the commit that removes the unused test and rebased with master again. Sorry for the incoveniences. Thanks!
